<?xml version="1.0" ?>
<robot name="darm_ros2" xmlns:xacro="http://www.ros.org/wiki/xacro">

<xacro:include filename="$(find darm_ros2)/description/materials.xacro" />
<xacro:include filename="$(find darm_ros2)/description/darm.trans" />
<xacro:include filename="$(find darm_ros2)/description/ros2_control.xacro" />

    <!-- LINKS -->
	<link name="world" />

    <link name="base_link">
        <inertial>
            <origin xyz="-2.4821E-09 1.3166E-10 0.029745" rpy="0 0 0" />
            <mass value="3.9468" />
            <inertia ixx="0.043126" ixy="-1.7415E-11" ixz="2.6551E-10" iyy="0.043126" iyz="-3.9532E-12" izz="0.081135" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry> <mesh filename="package://darm_ros2/meshes/base_link.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry> <mesh filename="package://darm_ros2/meshes/base_link.stl" /> </geometry>
        </collision>
    </link>

    <link name="link_1">
        <inertial>
            <origin xyz="7.1591E-08 0.058438 -8.6894E-09" rpy="0 0 0" />
            <mass value="0.92957" />
            <inertia ixx="0.0022123" ixy="2.3769E-09" ixz="3.7302E-09" iyy="0.0050488" iyz="-4.2711E-10" izz="0.0055584" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/link_1.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/link_1.stl" /> </geometry>
        </collision>
    </link>

    <link name="link_2">
        <inertial>
            <origin xyz="9.624E-09 0.31697 1.3709E-10" rpy="0 0 0" />
            <mass value="0.59942" />
            <inertia ixx="0.0027984" ixy="-1.549E-11" ixz="-2.8495E-12" iyy="0.0011453" iyz="2.7668E-13" izz="0.0030477" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/link_2.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/link_2.stl" /> </geometry>
        </collision>
    </link>

    <link name="link_3">
        <inertial>
            <origin xyz="0.0026787 0.056368 -5.0002E-08" rpy="0 0 0" />
            <mass value="0.28797" />
            <inertia ixx="0.00071119" ixy="-1.6217E-06" ixz="1.3749E-10" iyy="0.00048901" iyz="6.5455E-10" izz="0.00059962" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/link_3.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/link_3.stl" /> </geometry>
        </collision>
    </link>

    <link name="forearm">
        <inertial>
            <origin xyz="-0.01112 0.60989 1.3852E-09" rpy="0 0 0" />
            <mass value="0.72618" />
            <inertia ixx="0.0060414" ixy="-0.00087072" ixz="-2.2061E-11" iyy="0.0018715" iyz="6.2388E-11" izz="0.0072794" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/forearm.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/forearm.stl" /> </geometry>
        </collision>
    </link>

    <link name="differential">
        <inertial>
            <origin xyz="1.0228E-10 0.81054 -1.803E-10" rpy="0 0 0" />
            <mass value="0.12487" />
            <inertia ixx="3.4148E-05" ixy="-1.1807E-13" ixz="1.5487E-13" iyy="0.00014464" iyz="1.2478E-13" izz="0.00014697" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/differential.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/differential.stl" /> </geometry>
        </collision>
    </link>

    <link name="gripper">
        <inertial>
            <origin xyz="1.47508677776788E-10 0.87862732529105 4.19167841612699E-10" rpy="0 0 0" />
            <mass value="0.227833460574031" />
            <inertia ixx="0.000187517505664128" ixy="-3.14443887981606E-12" ixz="-3.98883278994357E-11" iyy="0.000181158961918109" iyz="1.97659221684485E-13" izz="0.000227045470734226" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/gripper.stl" /> </geometry>
            <material name="orange"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/gripper.stl" /> </geometry>
        </collision>
    </link>

    <link name="finger_1">
        <inertial>
            <origin xyz="-0.0162149507250985 0.938840073230663 -0.00086445843187576" rpy="0 0 0" />
            <mass value="0.0333577018228372" />
            <inertia ixx="1.08379131084182E-05" ixy="-1.84217384725124E-07" ixz="-3.69353084704429E-07" iyy="4.94385971740682E-06" iyz="-7.64596058297643E-07" izz="1.18448488567153E-05" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/finger_1.stl" /> </geometry>
            <material name="rubber"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/finger_1.stl" /> </geometry>
        </collision>
    </link>

    <link name="finger_2">
        <inertial>
            <origin xyz="0.0162149507252072 0.938840073230661 0.000864458431773843" rpy="0 0 0" />
            <mass value="0.0333577018228331" />
            <inertia ixx="1.08379131084146E-05" ixy="1.84217384719479E-07" ixz="-3.69353084703627E-07" iyy="4.94385971740466E-06" iyz="7.64596058296907E-07" izz="1.18448488567099E-05" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/finger_2.stl" /> </geometry>
            <material name="rubber"/>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/finger_2.stl" /> </geometry>
        </collision>
    </link>

    <link name="dock">
        <inertial>
            <origin xyz="9.123968339512894e-08 0.02244543692492791 -0.21216314201498515" rpy="0.0 0.0 0.0" />
            <mass value="4.14283366579888" />
            <inertia ixx="0.003902882548391967" iyy="0.00782818367286213" izz="0.006179527994280975" ixy="-2.754251235646076e-09" iyz="0.00017015521062901904" ixz="6.810558796533437e-09" />
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/dock.stl" scale="0.001 0.001 0.001" /> </geometry>
            <material name="rubber"/>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
            <geometry> <mesh filename="package://darm_ros2/meshes/dock.stl" scale="0.001 0.001 0.001" /> </geometry>
        </collision>
    </link>

    <!-- JOINTS -->
    <joint name="base_joint" type="fixed">
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <parent link="world"/>
        <child link="base_link"/>        
    </joint>

    <joint name="joint_1_joint" type="revolute">
        <origin xyz="0 -0.14444 0" rpy="${pi} 0 0" />
        <parent link="base_link" />
        <child link="link_1" />
        <axis xyz="0 1 0"/>
        <limit lower="${-pi - 0.5}" upper="${pi + 0.5}" effort="20" velocity="2" />
        <dynamics friction="0.2" />
    </joint>

    <joint name="joint_2_joint" type="revolute">
        <origin xyz="0 -0.14444 0" rpy="0 0 0" />
        <parent link="link_1" />
        <child link="link_2" />
        <axis xyz="1 0 0"/>
        <limit lower="-0.001" upper="${226 * pi / 180}" effort="40" velocity="2" />
        <dynamics friction="0.2" />
    </joint>

    <joint name="joint_3_joint" type="revolute">
        <origin xyz="0 0.40708 0" rpy="0 0 0" />
        <parent link="link_2" />
        <child link="link_3" />
        <axis xyz="0 1 0" />
        <limit lower="${-pi - 0.5}" upper="${pi + 0.5}" effort="20" velocity="2" />
        <dynamics friction="0.2" />
    </joint>

    <joint name="forearm_joint" type="revolute">
        <origin xyz="0 -0.40708 0" rpy="0 0 0" />
        <parent link="link_3" />
        <child link="forearm" />
        <axis xyz="-1 0 0"/>
        <limit lower="-0.001" upper="${122.9 * pi / 180}" effort="20" velocity="2" />
        <dynamics friction="0.2" />
    </joint>

    <joint name="differential_joint" type="revolute">
        <origin xyz="0 0 0" rpy="0 0 0" />
        <parent link="forearm" />
        <child link="differential" />
        <axis xyz="-1 0 0"/>
        <limit lower="${-110 * pi / 180}" upper="${110 * pi / 180}" effort="8" velocity="2" />
        <dynamics friction="0.2" />
    </joint>

    <joint name="gripper_joint" type="revolute">
        <origin xyz="0.052776 0.021377 0.003731" rpy="0 0 0"/>
        <parent link="differential" />
        <child link="gripper" />
        <axis xyz="0 -0.985109 -0.171929"/>
        <limit lower="-3.141593" upper="3.141593" effort="8" velocity="2" />
        <dynamics friction="0.2" />
    </joint>

    <joint name="finger_1_joint" type="prismatic">
        <origin xyz="0.0265 0 0" rpy="0 0 0" />
        <parent link="gripper" />
        <child link="finger_1" />
        <axis xyz="1 0 0"/>
        <limit lower="-0.001" upper="0.025" effort="2" velocity="0.5" />
        <dynamics friction="0.1" />
    </joint>

    <joint name="finger_2_joint" type="prismatic">
        <origin xyz="-0.0265 0 0" rpy="0 0 0" />
        <parent link="gripper" />
        <child link="finger_2" />
        <axis xyz="1 0 0"/>
        <limit lower="-0.001" upper="0.025" effort="2" velocity="0.5" />
        <dynamics friction="0.1" />
    </joint>

    <joint name="dock_joint" type="fixed">
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<parent link="base_link"/>
		<child link="dock"/>
	</joint>

    <!-- Plugins -->
    <gazebo>
        <plugin name="gz::sim::systems::JointStatePublisher" filename="gz-sim-joint-state-publisher-system"/>
    </gazebo>

</robot>